### ISSUE TITLE
[RyuJIT] Add support for bswap

### ISSUE THREAD

**Issue by erozenfeld** [dotnet/coreclr#14122]
The jit currently doesn't emit bswap instructions. We could add recognition of byte swap pattern to morph (similar to what we do for rotation).

This came up in https://github.com/dotnet/corefxlab/pull/1772, e.g.,
```
public static uint ReverseEndianness(uint value)
{
    return (value & 0x000000FFU) << 24 |
               (value & 0x0000FF00U) << 8 |
               (value & 0x00FF0000U) >> 8 |
               (value & 0xFF000000U) >> 24;
}
```

category:cq
theme:basic-cq
skill-level:intermediate
cost:small

**Comment by mikedn** [comment]
This should really be an intrinsic in the `BitConverter` class. The amount of IR to pattern match is pretty large and the code itself is pretty ugly to be duplicated all other the place. The only reason people write such code is that in more than 15 years of .NET nobody bothered to add some convenient endian conversion methods somewhere.

**Comment by ahsonkhan** [comment]
> The only reason people write such code is that in more than 15 years of .NET nobody bothered to add some convenient endian conversion methods somewhere.

We are trying do add some convenient endian conversion methods :)
See: https://github.com/dotnet/corefxlab/blob/master/src/System.Binary/System/Binary/BufferReader_BE.cs#L13

**Comment by mikedn** [comment]
> We are trying do add some convenient endian conversion methods :)

Convenient conversion method means something like `int ByteSwap(int)` or something more fancy like `int HostToBigEndian(int)`, `int HostToLittleEndian(int)` etc. In other words, methods that can be used in any context and are not associated in any way with a specific representation of a sequence of bytes (`byte[]`, `byte*`, `Span<byte>` etc.).

**Comment by benaadams** [comment]
> or something more fancy like int HostToBigEndian(int), int HostToLittleEndian(int) etc.

`HostToBigEndian` and `BigEndianToHost` are 

`IPAddress.NetworkToHostOrder` / `IPAddress.HostToNetworkOrder` `Int16`, `Int32`, `Int64`

Though `BitConverter` would be a more sensible place and still need `HostToLittleEndian`

**Comment by mikedn** [comment]
> HostToBigEndian and BigEndianToHost are IPAddress.NetworkToHostOrder / IPAddress.HostToNetworkOrder Int16, Int32, Int64

No, they are not. Host order is one thing, network order is another thing and big endian is yet another thing. Of course, they is some overlap but when you write code that deals with the big endian files you don't want to involve HostToNetworkOrder just because the network order happens to be big endian.

> Although sometimes you really just want to swap the bytes :)

Indeed, as far as intrinsics are concerned `ByteSwap` is enough. The rest can be built on top of `ByteSwap` and `BitConverter.IsLittleEndian`.

**Comment by GrabYourPitchforks** [comment]
I came up with a _very rough_ prototype of this implementation as part of learning how the JIT worked. See https://github.com/GrabYourPitchforks/coreclr/commit/2f80e64b1c30162e3f6f18c0afd15296a45a5614. It works in the limited test cases I've tried. Hopefully somebody can run with it and see it through to production.

**Comment by khellang** [comment]
Should this've been closed with #18398?

**Comment by GrabYourPitchforks** [comment]
Yup!
